docs/notebooks/flax_basics.ipynb

```diff
@@ -0,0 +1,911 @@
+{
```

Comment:
<p>In terms of structure, why do first have <code>nn.Module</code>, then a lot of stuff that's really JAX and not Flax, then optimization, then serialization, and then we come back to <code>nn.Module</code> ?</p><p> </p><p>Consider having all <code>nn.Module</code> stuff together. Personally, I also feel that optimization &amp; serialization should come after this, but YMMV.</p><p> </p><p>Maybe somewhere here it's a good point to stress again that we imported <code>linen as nn</code>,  not to be confounded with the old API.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/flax/pull/542/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Reply:
<p>Happy to discuss the structure, and the one you suggest also makes sense. The rationale for doing it this way was that it was more natural as a first time user, meaning that you are adding complexities on step at a time. Here the structure is more as a learning process, and I think that before wanting to define your own modules, you might want to start by understanding how to import, use, train models provided by others. After that you might be interested in defining your own models.</p><p> </p><p>I shared the notebooks with people who know nothing about Jax and Flax, and I'm interested in what they'll think, but my feeling was that this way was a bit more "progressive". We should get more opinion on this to see what's the best way.</p>

Reply:
Yes it would be great to get some unbiased input from first time users!

It would also be interesting to know what percentage of new Flax users are also new to JAX, vs. users who have already used JAX and now want to learn Flax.